### PR TITLE
 POC for integration API to decouple from OT

### PR DESCRIPTION
**Pull request by tylerbenson**
Won’t compile since most things aren’t migrated over.  The goal would be to work through a Strawman API that can be applied to the instrumentation, ideally without any changes to the public/surface api.  Once that's in place, it becomes easier to work out the changes needed for the underlying tracer.

(focus on the last commit: https://github.com/DataDog/dd-trace-java/commit/ff5d4cee8799f843c0cf94b9ac349a3d6a2314a5)

Thoughts?

[skip ci]

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/AgentTracer.java [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+  public static void registerIfAbsent(final TracerAPI trace) {
+    provider.compareAndSet(DEFAULT, trace);
+  }
+
+  public static TracerAPI get() {
+    return provider.get();
+  }
+
+  public interface TracerAPI {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/AgentTracer.java]: This is the primary interface showing all the methods (probably missing some) that are needed for writing instrumentation.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/AgentTracer.java [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    AgentSpan startSpan(BaseDecorator decorator, AgentSpan.Context parent);
+
+    AgentScope activateSpan(AgentSpan span);
+
+    AgentSpan activeSpan();
+
+    Propagation propagate();
+  }
+
+  private static final class OpenTracing31 implements TracerAPI {
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/AgentTracer.java]: This is the primary bridge which converts the api calls into OpenTracing lingo.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/Propagation.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  <C> void inject(AgentSpan span, C carrier, Setter<C> setter);
+
+  interface Setter<C> {
+    void set(C carrier, String key, String value);
+  }
+
+  <C> AgentSpan.Context extract(C carrier, Getter<C> getter);
+
+  interface Getter<C> {
+    List<String> keys(C carrier);
```
Comment (tylerbenson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/instrumentation/api/Propagation.java]: Originally I didn't include this method, but it's necessary to support arbitrary headers (with things like ot-baggage that is based on a prefix).

**Comment by dougqh** [comment]
So far, it looks good to me -- with the exception of thinking we should we have our own constants class.

I see that most the work is mechanical replacement.  It might be a little easier to review if each mechanical replacement was its own commit, so I can look at each change in isolation.

**Comment by tylerbenson** [comment]
Closing in favor of #1020 and subsequent work.

**Pull request closed without merging**